File: 90-subsets-ii/90-subsets-ii.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        ans=new ArrayList<>();
        Arrays.sort(nums);
        dfsSubset(nums, 0, new ArrayList<>());
        return ans;
    }
    void dfsSubset(int[] nums, int i, List<Integer> list){
        if(i==nums.length){
            ans.add(new ArrayList<>(list));
            return;
        }
        list.add(nums[i]);
        dfsSubset(nums, i+1, list);
        list.remove(list.size()-1);
        int j=i;
        while(j<nums.length && nums[j]==nums[i]) j++;
        dfsSubset(nums, j, list);
    }
}

File: 90-subsets-ii/90-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans=[]
        nums.sort()
        
        def dfs(i, subset):
            if i==len(nums):
                ans.append(subset.copy())
                return
            subset.append(nums[i])
            dfs(i+1, subset)
            subset.pop()
            j=i
            while j<len(nums) and nums[i]==nums[j]:
                j+=1
            dfs(j, subset)
            
        dfs(0, [])
        return ans
        